File: app/lib/domain/value_objects/contact_message.dart
final class ContactMessage {
  const ContactMessage._({
    required this.name,
    required this.email,
    required this.body,
  });

  factory ContactMessage({
    String? name,
    required String email,
    required String body,
  }) {
    final trimmedName = (name ?? '').trim();
    final trimmedEmail = email.trim();
    final trimmedBody = body.trim();

    if (trimmedName.length > maxNameLength) {
      throw ArgumentError.value(
        name,
        'name',
        'must be $maxNameLength characters or fewer '
            '(got ${trimmedName.length})',
      );
    }
    if (trimmedEmail.length > maxEmailLength ||
        !_emailPattern.hasMatch(trimmedEmail)) {
      throw ArgumentError.value(email, 'email', 'must be a valid address');
    }
    if (trimmedBody.length < minBodyLength) {
      throw ArgumentError.value(
        body,
        'message',
        'must be at least $minBodyLength characters',
      );
    }
    if (trimmedBody.length > maxBodyLength) {
      throw ArgumentError.value(
        body,
        'message',
        'must be $maxBodyLength characters or fewer '
            '(got ${trimmedBody.length})',
      );
    }

    return ContactMessage._(
      name: trimmedName.isEmpty ? null : trimmedName,
      email: trimmedEmail,
      body: trimmedBody,
    );
  }

  static const maxNameLength = 80;
  static const maxEmailLength = 254;
  static const minBodyLength = 10;
  static const maxBodyLength = 4000;

  static final _emailPattern = RegExp(
    r'^[^\s<>"@]+@[^\s<>"@.]+(\.[^\s<>"@.]+)+$',
  );

  final String? name;
  final String email;
  final String body;

  @override
  bool operator ==(Object other) =>
      other is ContactMessage &&
      other.name == name &&
      other.email == email &&
      other.body == body;

  @override
  int get hashCode => Object.hash(name, email, body);

  @override
  String toString() => 'ContactMessage(${name ?? '(no name)'} <$email>)';
}

File: web/src/domain/value-objects/contact-message.ts
import { type Failure, FailureField, type InvalidInputFailure, invalidInput } from "../failures/failure";

export const CONTACT_ROUTE = "/api/contact";

export const MAX_CONTACT_NAME_LENGTH = 80;
export const MAX_CONTACT_EMAIL_LENGTH = 254;
export const MIN_CONTACT_BODY_LENGTH = 10;
export const MAX_CONTACT_BODY_LENGTH = 4000;

const CONTACT_EMAIL_REGEX = /^[^\s<>"@]+@[^\s<>"@.]+(?:\.[^\s<>"@.]+)+$/;

export interface ContactMessage {
	readonly _tag: "ContactMessage";
	readonly name: string | null;
	readonly email: string;
	readonly body: string;
}

export const validateContactName = (name: string | null | undefined): InvalidInputFailure | null => {
	if ((name ?? "").trim().length > MAX_CONTACT_NAME_LENGTH) {
		return invalidInput({
			field: FailureField.Name,
			message: `Name must be ${MAX_CONTACT_NAME_LENGTH} characters or fewer`,
		});
	}
	return null;
};

export const validateContactEmail = (email: string): InvalidInputFailure | null => {
	const trimmed = email.trim();
	if (trimmed === "") {
		return invalidInput({ field: FailureField.Email, message: "Enter your email address" });
	}
	if (trimmed.length > MAX_CONTACT_EMAIL_LENGTH || !CONTACT_EMAIL_REGEX.test(trimmed)) {
		return invalidInput({ field: FailureField.Email, message: "Enter a valid email address" });
	}
	return null;
};

export const validateContactBody = (body: string): InvalidInputFailure | null => {
	const trimmed = body.trim();
	if (trimmed === "") {
		return invalidInput({ field: FailureField.Message, message: "Write a message" });
	}
	if (trimmed.length < MIN_CONTACT_BODY_LENGTH) {
		return invalidInput({
			field: FailureField.Message,
			message: `Message must be at least ${MIN_CONTACT_BODY_LENGTH} characters`,
		});
	}
	if (trimmed.length > MAX_CONTACT_BODY_LENGTH) {
		return invalidInput({
			field: FailureField.Message,
			message: `Message must be ${MAX_CONTACT_BODY_LENGTH} characters or fewer`,
		});
	}
	return null;
};

export interface ParseContactMessageParams {
	name?: string | null;
	email: string;
	body: string;
}

export const parseContactMessage = ({ name, email, body }: ParseContactMessageParams): ContactMessage | Failure => {
	const rejected = validateContactName(name) ?? validateContactEmail(email) ?? validateContactBody(body);
	if (rejected) return rejected;

	const trimmedName = (name ?? "").trim();
	return {
		_tag: "ContactMessage",
		name: trimmedName === "" ? null : trimmedName,
		email: email.trim(),
		body: body.trim(),
	};
};

export const isContactMessage = (value: unknown): value is ContactMessage =>
	typeof value === "object" && value !== null && (value as { _tag?: unknown })._tag === "ContactMessage";
